vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1055 @@
+// Low level SPI definitions
+//
+// Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+//          Haixu Cui <quic_haixcui@quicinc.com>
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use std::{
+    convert::From,
+    fs::{File, OpenOptions},
+    os::unix::io::AsRawFd,
+    ptr,
+};
+
+use thiserror::Error as ThisError;
+use vmm_sys_util::errno::Error as IoError;
+use vmm_sys_util::ioctl::{ioctl_with_mut_ptr, ioctl_with_mut_ref, ioctl_with_ref};
+
+use crate::{vhu_spi::VirtioSpiConfig, virtio_spi::*};
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, PartialEq, ThisError)]
+/// Errors related to low level spi helpers
+pub(crate) enum Error {
+    #[error("Ioctl command failed for {0} operation: {1}")]
+    IoctlFailure(&'static str, IoError),
+    #[error("Failed to open spi controller")]
+    DeviceOpenFailed,
+}
+
+/// Describes a single SPI transfer
+#[derive(Debug)]
+#[repr(C)]
+pub(crate) struct SpiIocTransfer {
+    /// Holds pointer to userspace buffer with transmit data, or null
+    tx_buf: u64,
+    /// Holds pointer to userspace buffer for receive data, or null.
+    rx_buf: u64,
+    /// Length of tx and rx buffers, in bytes.
+    len: u32,
+    /// Temporary override of the device's bitrate.
+    speed_hz: u32,
+    /// If nonzero, how long to delay after the last bit transfer
+    /// before optionally deselecting the device before the next transfer.
+    delay_usecs: u16,
+    /// Temporary override of the device's wordsize.
+    bits_per_word: u8,
+    /// True to deselect device before starting the next transfer.
+    cs_change: u8,
+    /// Number of bits used for writing.
+    tx_nbits: u8,
+    /// Number of bits used for reading.
+    rx_nbits: u8,
+    /// If nonzero, how long to wait between words within one
+    /// transfer. This property needs explicit support in the SPI controller,
+    /// otherwise it is silently ignored
+    word_delay_usecs: u8,
+    pad: u8,
+}
+
+/// SPI definitions
+pub(crate) struct SpiTransReq {
+    pub tx_buf: Vec<u8>,
+    pub rx_buf: Vec<u8>,
+    pub trans_len: u32,
+    pub speed_hz: u32,
+    pub mode: u32,
+    pub delay_usecs: u16,
+    pub bits_per_word: u8,
+    pub cs_change: u8,
+    pub tx_nbits: u8,
+    pub rx_nbits: u8,
+    pub word_delay_usecs: u8,
+    pub cs_id: u8,
+}
+
+/// Trait that represents a SPI Device.
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the SPI driver so that we can test the SPI
+/// functionality without the need of a physical device.
+pub(crate) trait SpiDevice {
+    // Open the device specified by the controller path.
+    fn open(path: &str) -> Result<Self>
```

Comment:
```suggestion
    fn open(path: &Path) -> Result<Self>
```

Reply:
Now it is:
`fn open(path: &PathBuf) -> Result<Self>`